src/messages.rs

```diff
@@ -46,27 +46,29 @@ macro_rules! impl_message_reader {
             type Item = $output;
 
             fn next(&mut self) -> Option<Self::Item> {
-                let mut first_null = None;
-                for (i, b) in self.byte_queue.iter().enumerate() {
-                    if *b == '\0' as u8 {
-                        first_null = Some(i);
-                        break;
-                    }
+                // We need two bytes for the length
+                if self.byte_queue.len() < 2 {
+                    return None;
                 }
 
-                let msg_bytes = match first_null {
-                    None => return None,
-                    Some(i) => self.byte_queue.drain(0..i)
-                }.collect::<Vec<u8>>();
-                self.byte_queue.pop_front();
+                let length = u16::from_be_bytes([
+                    self.byte_queue[0],
+                    self.byte_queue[1],
+                ]) as usize;
 
-                let as_str = String::from_utf8_lossy(&msg_bytes);
+                // We will not read a message until a complete message has been
+                // received
+                if self.byte_queue.len() < 2 + length {
+                    return None;
+                }
+
+                self.byte_queue.pop_front().unwrap();
```

Comment:
Why is this done?

Reply:
The length of the rest of the message is stored in the first two bytes

Reply:
But why are the bytes popped and not looked at?

Reply:
They are looked at. When we read the length out. But we don't want to pop them until we are sure that our entire message has arrived so we can read it all at once.